node/adapter/docker_restart.py

```diff
@@ -0,0 +1,125 @@
+import os
+import time
+import sh
+import fileinput
+import sys
+REAL_SOCK = "/var/run/docker.real.sock"
+POWERSTRIP_SOCK = "/var/run/docker.sock"
+
+
+def _replace_all(filename, search, replace):
+    for line in fileinput.input(filename, inplace=1):
+        if search in line:
+            line = line.replace(search, replace)
+        sys.stdout.write(line)
+
+def create_restarter():
+    """
+    Detect what init system is being used and return the appropriate handler.
+    :return: A "restarter" object.
+    """
+    if os.path.exists(SystemdRestarter.DOCKER_SYSTEMD_SERVICE):
+        return SystemdRestarter()
+    elif os.path.exists(UpstartRestarter.DOCKER_DEFAULT_FILENAME):
+        return UpstartRestarter()
+    else:
+        return NullRestarter()
+
+
+class NullRestarter():
+    def is_using_alternative_socket(self):
+        return False
+    def restart_docker_with_alternative_unix_socket(self):
+        print "Unsupported"
+    def restart_docker_without_alternative_unix_socket(self):
+        print "Unsupported"
+
+
+def _clean_socks():
+    if os.path.exists(REAL_SOCK):
+        os.remove(REAL_SOCK)
+    if os.path.exists(POWERSTRIP_SOCK):
+        os.remove(POWERSTRIP_SOCK)
+
+
+class SystemdRestarter():
+    def __init__(self):
+        pass
+
+    DOCKER_SYSTEMD_SERVICE = "/usr/lib/systemd/system/docker.service"
+    SYSTEMD_DEFAULT = "ExecStart=/usr/bin/docker -d $OPTIONS \\"
+    SYSTEMD_MODIFIED = "ExecStart=/usr/bin/docker -d $OPTIONS " \
+                       "-H unix://%s \\" % REAL_SOCK
+
+    def _clean_restart_docker(self, sock_to_wait_on):
+        _clean_socks()
+        systemctl = sh.Command._create("systemctl")
+        systemctl("daemon-reload")
+        systemctl("restart", "docker.service")
+
+        # Wait for docker to create the socket
+        while not os.path.exists(sock_to_wait_on):
```

Comment:
We probably want to timeout with an error message if we wait too long.  30 seconds?